File: mnpbem-mie/src/mnpbem_mie/rayleigh.py
"""Small-particle (Rayleigh) scattering formulas for quick spectral estimates."""

from __future__ import annotations

import numpy as np


def rayleigh_polarizability(
    radius_nm: float,
    eps_particle: np.ndarray | complex,
    eps_medium: np.ndarray | complex,
) -> np.ndarray:
    """Compute electrostatic polarizability alpha = 4*pi*a^3*(ep-em)/(ep+2*em)."""
    a = float(radius_nm)
    ep = np.asarray(eps_particle, dtype=complex)
    em = np.asarray(eps_medium, dtype=complex)
    return 4.0 * np.pi * a**3 * (ep - em) / (ep + 2.0 * em)


def rayleigh_cross_sections(
    wavelength_nm: np.ndarray | float,
    radius_nm: float,
    eps_particle: np.ndarray | complex,
    eps_medium: np.ndarray | complex,
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Return (C_ext, C_sca, C_abs) using Rayleigh-limit closed forms."""
    wavelength = np.asarray(wavelength_nm, dtype=float)
    em = np.asarray(eps_medium, dtype=complex)
    k = 2.0 * np.pi * np.sqrt(em) / wavelength

    alpha = rayleigh_polarizability(radius_nm, eps_particle, eps_medium)

    c_ext = k * np.imag(alpha)
    c_sca = (np.abs(k) ** 4) * np.abs(alpha) ** 2 / (6.0 * np.pi)
    c_abs = c_ext - c_sca
    return c_ext.real, c_sca.real, c_abs.real

File: npm-packages/packages/mnp-mie/src/rayleigh.js
import { abs, add, complex, div, fromReal, imag, mul, sub } from "./complex.js";

function mapInput(values, fn) {
  if (Array.isArray(values)) {
    return values.map(fn);
  }
  return fn(values);
}

export function rayleighPolarizability(radiusNm, epsParticle, epsMedium) {
  const a = Number(radiusNm);
  const top = sub(epsParticle, epsMedium);
  const bottom = add(epsParticle, mul(fromReal(2), epsMedium));
  return mul(fromReal(4 * Math.PI * a ** 3), div(top, bottom));
}

export function rayleighCrossSections(wavelengthNm, radiusNm, epsParticle, epsMedium) {
  const solveOne = (wl, ep, em) => {
    const kFactor = (2 * Math.PI * Math.sqrt(em.re)) / Number(wl);
    const k = fromReal(kFactor);
    const alpha = rayleighPolarizability(radiusNm, ep, em);
    const cExt = k.re * imag(alpha);
    const cSca = (Math.abs(k.re) ** 4 * abs(alpha) ** 2) / (6 * Math.PI);
    const cAbs = cExt - cSca;
    return { cExt, cSca, cAbs };
  };

  if (Array.isArray(wavelengthNm)) {
    return wavelengthNm.map((wl, i) => solveOne(wl, epsParticle[i], epsMedium[i]));
  }
  return solveOne(wavelengthNm, epsParticle, epsMedium);
}

export { complex };
